README.md

```diff
@@ -1,7 +1,214 @@
-#TBA...
+# ApiProvider
 
-## Features and bugs
+### Clean and boilerplate free network requests with only a few lines of code.
 
-Please file feature requests and bugs at the [issue tracker][tracker].
+```dart
+  final request = Request(url: Path("https://someapi/user"), httpMethod: HttpMethod.GET);
 
-[tracker]: https://github.com/tikkrapp/api_provider/issues
+  final response = await apiProvider.request<User>(request: request);
+  print("fetched user :${response.body.name}");
+```
+
+### Features:
+
+- Typesafe interface on `Request` and `Response` objects.
+- Powerful `Interceptors`
+- Testing / Mocking out of the Box!
+
+
+
+### Installing
+
+Add the following snippet into your `pubspec.yaml`. (Release on pub will happen when the api is stable)
+
+```yaml
+api_provider:
+    git:
+      url: git://github.com/tikkrapp/api_provider.git
+```
+
+
+
+
+
+### Serializing Requests and Deserializing Responses
+
+Instead of providing a serializer when calling  `request`, all the serialisers are stored in a `SerializeContainer` .The `ApiProvider` will get the stored  `Serializable`'s and `Deserializable` 's that are needed when you call `request`.
+
+⚠️  Important:  `T`  of `request` is used to get the needed serializer, thats why it should never be  `dynamic`.  ⚠️ 
+
+We highly recomment to disable `implicit-dynamic`  and `implicit-casts`.
+
+```yml
+analyzer:
+  strong-mode:
+    implicit-casts: false
+    implicit-dynamic: false
+```
+
+
+
+Specify your `SerializeContainer` in the constructor.
+
+```dart
+final apiProvider = HttpClientApiProvider(
+    // use IOClient for flutter apps
+    // use BrowserClient for dartJs
+    httpClient: client,
+    container: SerializerContainer()
+      ..insertEncoder(UserEncoder())
+      ..insertDecoder(UserDecoder()),
+);
+
+// responsible for decoding a user
+// similar to a "toMap" method.
+class UserEncoder implements Deserializable<Map, User> {
+  @override
+  Map to(User input) => <String, dynamic>{
+        'email': input.email,
+        'name': input.name,
+      };
+}
+
+// responsible for encoding a user
+// similar to a "fromMap" method.
+class UserDecoder implements Serializable<Map, User> {
+  @override
+  User from(Map input) => User(
+    email: input['email'] as String,
+    name: input['name'] as String,
+  );
+}
+```
+
+
+
+### Interceptors
+
+Interceptors give you the ability to transform, change or listen to request or response calls.
+
+This could be something simply like adding an Authorization Header or something more complex like catching and handling a specific error from your api.
+
+1. #### Start by implementing `Interceptor`. 
+
+```dart
+class MyInterceptor implements Interceptor {
+  @override
+  RequestHandler<T> intercept<T>(RequestHandler<T> handler) => (request) async {
+        // adjust the request here using request.copyWith
```

Comment:
Writing "manipulate" instead of "adjust" seems to be a better choice from my point of view.